### PR TITLE
Add iptable rules to deny VNC access on Compute hosts (bnc #938200)

### PR DESCRIPTION
**Pull request by dguitarbite**
VNC ports are accessible from the management network. This patch
implements iptable rules which will deny access to these ports to all
but the novnc proxy service (nova-controller).

### CODE REVIEW

File: chef/cookbooks/nova/recipes/compute.rb [hunk 1]
```diff
@@ -418,3 +418,12 @@ def set_boot_kernel_and_trigger_reboot(flavor="default")
 execute "trigger-nova-own-az-config" do
   command "true"
 end
+
+# Set iptables rules for blocking VNC Access for all but the nova-controller node.
+bash "vnc_access" do
+  code <<-EOH
+    touch /tmp/hello_pranav
```
Comment (mjura) [inline chef/cookbooks/nova/recipes/compute.rb]: We don't need this

File: chef/cookbooks/nova/recipes/compute.rb [hunk 2]
```diff
@@ -418,3 +418,12 @@ def set_boot_kernel_and_trigger_reboot(flavor="default")
...
+    iptables -I INPUT -p tcp --match multiport --dports 5900:15900 -j DROP
+    iptables -I INPUT -s #{nova_controller_ip} -p tcp --match multiport --dports 5900:15900 -j ACCEPT
```
Comment (mjura) [inline chef/cookbooks/nova/recipes/compute.rb]: This two iptables rules can be replaced with one rule

**Comment by mjura** [comment]
We should consider to use for this task /etc/host.deny TCP wrapper instead iptables which is managed by OpenStack on compute nodes.

File: chef/cookbooks/nova/recipes/compute.rb (hunk 1 above)
Comment (dguitarbite) [inline chef/cookbooks/nova/recipes/compute.rb]: Removing this. It sneaked into this patch.

File: chef/cookbooks/nova/recipes/compute.rb (hunk 2 above)
Comment (dguitarbite) [inline chef/cookbooks/nova/recipes/compute.rb]: I am taking a look into this. If I find a good way to do this in one line, ill update it but as far as I could search for a single line rule, people have mentioned and preferred using multiple lines for such tasks.

**Comment by dguitarbite** [comment]
@mjura I have tried to use host.deny/allow. There are two issues here:

1. host.deny/allow does not allow us to select port range.
2. It has some similar limitations: The VNC ports are opened by qemu service that gives two more challenges:
a. Blocking qemu may break other services (live migration etc.) esp. since we cannot block specific ports.
b. Qemu by default in the Cloud product is not compiled with tcp flag, hence it bypasses hosts.allow/deny rules completely.

File: chef/cookbooks/nova/recipes/compute.rb (hunk 1 above)
Comment (ovisan) [inline chef/cookbooks/nova/recipes/compute.rb]: 1. hello_pranav ? please change the file name to something professional
2. delete the file after execution

File: chef/cookbooks/nova/recipes/compute.rb [hunk 3]
```diff
@@ -418,3 +418,12 @@ def set_boot_kernel_and_trigger_reboot(flavor="default")
 execute "trigger-nova-own-az-config" do
   command "true"
 end
+
+# Set iptables rules for blocking VNC Access for all but the nova-controller node.
+bash "vnc_access" do
```
Comment (tboerger) [inline chef/cookbooks/nova/recipes/compute.rb]: This is a pretty bad resource name. This name have to be unique within chef.

**Comment by tboerger** [comment]
Gating fails, bad naming, dummy files... -1

File: chef/cookbooks/nova/recipes/compute.rb (hunk 3 above)
Comment (mjura) [inline chef/cookbooks/nova/recipes/compute.rb]: We would like to run system commands using Chef calls with escaping.

File: chef/cookbooks/nova/recipes/compute.rb [hunk 4]
```diff
@@ -418,3 +418,12 @@ def set_boot_kernel_and_trigger_reboot(flavor="default")
...
+  code <<-EOH
+    touch /tmp/hello_pranav
+    iptables -I INPUT -p tcp --match multiport --dports 5900:15900 -j DROP
```
Comment (MaximilianMeister) [inline chef/cookbooks/nova/recipes/compute.rb]: i would decide either to use the full name of the options or the abbreviation, mixed like this it looks a bit weird

**Comment by rsalevsky** [comment]
Commit message is wrong we use `bsc#XXXXXX` without a space. Maybe add the complete URL to the commit description.

File: chef/cookbooks/nova/recipes/compute.rb [hunk 5]
```diff
@@ -418,3 +418,23 @@ def set_boot_kernel_and_trigger_reboot(flavor="default")
 execute "trigger-nova-own-az-config" do
   command "true"
 end
+
+admin_net = node[:network][:networks][:admin]
+admin_subnet = admin_net[:subnet] + "/" + admin_net[:netmask]
+
+# Set iptables rules for blocking VNC Access for all but the nova-controller node.
+bash "vncblock" do
```
Comment (tboerger) [inline chef/cookbooks/nova/recipes/compute.rb]: Still bad resource naming, if some other cookbook gives the same name to another resource the resource will just be reused. So avoid that!

Comment (dguitarbite) [inline chef/cookbooks/nova/recipes/compute.rb]: Do you suggest a name here? I do not know how exactly to name this resource.

File: chef/cookbooks/nova/recipes/compute.rb [hunk 6]
```diff
@@ -418,3 +418,23 @@ def set_boot_kernel_and_trigger_reboot(flavor="default")
...
+  code <<-EOH
+    iptables -N VNCBLOCK
+    # Ceph uses range 6800-7300~ which potentially conflicts with the VNC ports range.
+    # Setting the upper limit to 6799 to avoid blocking Ceph traffic for the time being.
+    # FIXME (psalunke): Update it to 15900 or more after changing the default Ceph ports.
+    # Also consider open ports on controller node in case of dual Controller/Compute node
+    # multiport --dports 8773:8779
+    iptables -D INPUT -j VNCBLOCK
+    iptables -I INPUT -j VNCBLOCK
+    iptables -F VNCBLOCK
```
Comment (mjura) [inline chef/cookbooks/nova/recipes/compute.rb]: Let's understand this properly:
- first you create a chain, 
- after this you delete chain with target, 
- then you create this same chain with a target
- and at the end you are flushing all rules to add your rule
For me it is insane

File: chef/cookbooks/nova/recipes/compute.rb [hunk 7]
```diff
@@ -418,3 +418,23 @@ def set_boot_kernel_and_trigger_reboot(flavor="default")
...
+
+# Set iptables rules for blocking VNC Access for all but the nova-controller node.
+bash "vncblock" do
+  code <<-EOH
+    iptables -N VNCBLOCK
+    # Ceph uses range 6800-7300~ which potentially conflicts with the VNC ports range.
+    # Setting the upper limit to 6799 to avoid blocking Ceph traffic for the time being.
+    # FIXME (psalunke): Update it to 15900 or more after changing the default Ceph ports.
+    # Also consider open ports on controller node in case of dual Controller/Compute node
+    # multiport --dports 8773:8779
```
Comment (mjura) [inline chef/cookbooks/nova/recipes/compute.rb]: On port 5984 couchdb is also listening

File: chef/cookbooks/nova/recipes/compute.rb (hunk 5 above)
Comment (mjura) [inline chef/cookbooks/nova/recipes/compute.rb]: I will repeat it once again, for system commands we would like to use Chef calls

Comment (tboerger) [inline chef/cookbooks/nova/recipes/compute.rb]: What about `nova_compute_vncblock`?

File: chef/cookbooks/nova/recipes/compute.rb [hunk 8]
```diff
@@ -418,3 +418,23 @@ def set_boot_kernel_and_trigger_reboot(flavor="default")
 execute "trigger-nova-own-az-config" do
   command "true"
 end
+
+admin_net = node[:network][:networks][:admin]
+admin_subnet = admin_net[:subnet] + "/" + admin_net[:netmask]
+
+# Set iptables rules for blocking VNC Access for all but the nova-controller node.
```
Comment (mjura) [inline chef/cookbooks/nova/recipes/compute.rb]: In comment you mention about `nova-controller` but you are allowing for whole admin network, please make it strict

**Comment by mjura** [comment]
If you want to delete and recreate iptable chains every each run of chef-client, then I would prefer to check on which ports nova services are listening and use only these specific ports. This was proposed by @dirkmueller during discussion about this card.

**Comment by dguitarbite** [comment]
WIP: Please do not merge this.

File: chef/cookbooks/nova/recipes/compute.rb [hunk 9]
```diff
@@ -285,16 +285,22 @@ def set_boot_kernel_and_trigger_reboot(flavor="default")
 
 env_filter = " AND nova_config_environment:#{node[:nova][:config][:environment]}"
 nova_controller = search(:node, "roles:nova-multi-controller#{env_filter}")
+nova_controller_ips = Array.new
```
Comment (tboerger) [inline chef/cookbooks/nova/recipes/compute.rb]: Just `nova_controller_ips = []`

**Comment by tboerger** [comment]
I can't believe that we need such complicated rules...

**Comment by dguitarbite** [comment]
You can merge this code now. @tboerger I agree, its too complex and we are doing some really not-kool-stuff in there :(.

**Comment by toabctl** [comment]
Hm. No PR testing here? Why?

**Comment by dguitarbite** [comment]
Just meant to say that its no longer WIP.

**Comment by tboerger** [comment]
IMHO it looks too complicated. Somebody with enough iptabpes knowledge needs to give some hints, please!

**Comment by ovisan** [comment]
we are doing? who is we?

File: chef/cookbooks/nova/recipes/compute.rb [hunk 10]
```diff
@@ -418,3 +424,36 @@ def set_boot_kernel_and_trigger_reboot(flavor="default")
...
+bash "nova_compute_vncblock_reject_all" do
+  code <<-EOH
+    iptables -D INPUT -j VNCBLOCK -p tcp --match multiport --dports 5900:15900
+    iptables -I INPUT -j VNCBLOCK -p tcp --match multiport --dports 5900:15900
+    iptables -F VNCBLOCK
+    iptables -I VNCBLOCK -p tcp \
+        -m connbytes --connbytes 0:1024 \
+        --connbytes-dir both --connbytes-mode bytes \
+        -m state --state ESTABLISHED \
+        -m u32 --u32 "0>>22&0x3C@ 12>>26&0x3C@ 0=0x52464220" \
```
Comment (ovisan) [inline chef/cookbooks/nova/recipes/compute.rb]: Can you please explain what this cryptic line does?

**Pull request closed without merging**